File: src/jirb_integration.rb
require 'java'
require 'jruby'
require 'irb'
require 'irb/completion'

# adapted from the jruby distro: bin/jirb_swing

class FrameBringer
	include java.lang.Runnable
	
	def run
		text = javax.swing.JTextPane.new
		text.font = find_font('Monospaced', java.awt.Font::PLAIN, 14, 'Monaco', 'Andale Mono')
		text.margin = java.awt.Insets.new(8,8,8,8)
		text.caret_color = java.awt.Color.new(0xa4, 0x00, 0x00)
		text.background = java.awt.Color.new(0xf2, 0xf2, 0xf2)
		text.foreground = java.awt.Color.new(0xa4, 0x00, 0x00)
		
		pane = javax.swing.JScrollPane.new
		pane.viewport_view = text
		
		frame = javax.swing.JFrame.new("JRuby v#{JRUBY_VERSION} IRB Console (tab will autocomplete)")
		frame.default_close_operation = javax.swing.JFrame::DO_NOTHING_ON_CLOSE
		frame.set_size(700, 600)
		frame.content_pane.add(pane)
		
		tar = org.jruby.demo.TextAreaReadline.new(text,
			  " JRuby VST Plugin Console - running plugin instances are in array PLUGS \n\n")
		
		JRuby.objectspace = false # useful for code completion, but BIG performance hit! --> set to false
		
		tar.hook_into_runtime_with_streams(JRuby.runtime)
		
		# show frame
		frame.visible = true
	end
end

def find_font(otherwise, style, size, *families)
	avail_families = java.awt.GraphicsEnvironment.local_graphics_environment.available_font_family_names
	fontname = families.find(proc {otherwise}) { |name| avail_families.include? name }
	java.awt.Font.new(fontname, style, size)
end

# trick IRB to think that we are a terminal --> now uses :DEFAULT prompt instead of the :NULL prompt
def STDIN.tty?
	true
end

def display_jirb	
	# do swing stuff on the EDT (event dispatch thread)
	java.awt.EventQueue.invoke_later(FrameBringer.new())
	
	# or just on a separate thread
	#irb_thread = Thread.new {
	#	FrameBringer.new().run()
	#}
	
	# need to wait a little to ensure that "tar.hook_into_runtime_with_streams(JRuby.runtime)" is completet
	sleep(1)
		
	IRB.start
end

File: src/JIRBIntegration.java

import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Font;
import java.awt.GraphicsEnvironment;
import java.awt.Insets;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.Arrays;
 
import javax.swing.BorderFactory;
import javax.swing.JEditorPane;
import javax.swing.JFrame;
import javax.swing.JScrollPane;
import javax.swing.JTextPane;
 
import org.jruby.Ruby;
import org.jruby.RubyInstanceConfig;
import org.jruby.internal.runtime.ValueAccessor;
import org.jruby.demo.TextAreaReadline;
 
import jvst.wrapper.VSTPluginAdapter;
 
 
public class JIRBIntegration extends JFrame {
    public JIRBIntegration(String title) {
        super(title);
    }
 
    static Ruby startRuby(boolean attachJIRB) {
	
		if (attachJIRB) {
			VSTPluginAdapter.log("Attaching JIRB...");
			
			final JIRBIntegration console = new JIRBIntegration("JRuby IRB Console (tab will autocomplete)");
			
			console.getContentPane().setLayout(new BorderLayout());
			console.setSize(700, 600);
			
			//do nothing on close (especially, we do not want to terminate the JVM (this would be the default!))
			console.setDefaultCloseOperation(DO_NOTHING_ON_CLOSE);
			
			JEditorPane text = new JTextPane();
			
			text.setMargin(new Insets(8,8,8,8));
			text.setCaretColor(new Color(0xa4, 0x00, 0x00));
			text.setBackground(new Color(0xf2, 0xf2, 0xf2));
			text.setForeground(new Color(0xa4, 0x00, 0x00));
			Font font = console.findFont("Monospaced", Font.PLAIN, 14,
					new String[] {"Monaco", "Andale Mono"});
			
			text.setFont(font);
			JScrollPane pane = new JScrollPane();
			pane.setViewportView(text);
			pane.setBorder(BorderFactory.createLineBorder(Color.darkGray));
			console.getContentPane().add(pane);
			console.validate();
			
			final TextAreaReadline tar = new TextAreaReadline(text, " JRuby VST Plugin Console - running plugin instances are in array PLUGS \n\n");
			
			final RubyInstanceConfig config = new RubyInstanceConfig() {{
				setInput(tar.getInputStream());
				setOutput(new PrintStream(tar.getOutputStream()));
				setError(new PrintStream(tar.getOutputStream()));
				setObjectSpaceEnabled(false); // would be useful for code completion inside the IRB, but BIG PERFORMCE HIT!
				setArgv(new String[]{}); //no args
			}};
			
			final Ruby runtime = Ruby.newInstance(config);
			
			runtime.getGlobalVariables().defineReadonly("$$", new ValueAccessor(runtime.newFixnum(System.identityHashCode(runtime))));
			runtime.getLoadService().init(new ArrayList());
			
			tar.hookIntoRuntime(runtime);
			
			//why not run on the EDT???
			//@Thibaut, you may need to change this to run on the EDT (java.awt.EventQueue.invoke_later(FrameBringer.new()))
			Thread t2 = new Thread() {
				public void run() {
					console.setVisible(true);
					runtime.evalScriptlet("require 'irb'; require 'irb/completion'; IRB.start");
				}
			};
			t2.start();
			
			return runtime;
		}
		else {
			Ruby runtime = Ruby.getDefaultInstance();
			return runtime;
		}
    }
 
    private Font findFont(String otherwise, int style, int size, String[] families) {
        String[] fonts = GraphicsEnvironment.getLocalGraphicsEnvironment().getAvailableFontFamilyNames();
        Arrays.sort(fonts);
        Font font = null;
        for (int i = 0; i < families.length; i++) {
            if (Arrays.binarySearch(fonts, families[i]) >= 0) {
                font = new Font(families[i], style, size);
                break;
            }
        }
        if (font == null)
            font = new Font(otherwise, style, size);
        return font;
    }
 
    private static final long serialVersionUID = 374624297323217387L;
 
}